cf-monitord/mon_network.c

```diff
@@ -202,42 +218,288 @@ void MonNetworkGatherData(double *cf_this)
 {
     ResetNetworkData();
 
-    char comm[PATH_MAX + 3] = {0};
-    strncpy(comm, VNETSTAT[VSYSTEMHARDCLASS], (sizeof(comm) - 1));
+    Item *in[ATTR] = {0};
+    Item *out[ATTR] = {0};
 
-    if (!FileCanOpen(comm, "r"))
+#ifdef __linux__
+    if ((access("/proc/net/tcp", R_OK) != 0) || !GetNetworkDataFromProcNet(cf_this, in, out))
+#endif
     {
-        Log(LOG_LEVEL_VERBOSE,
-            "Cannot open '%s', aborting gathering of network data (monitoring)",
-            comm);
-        return;
-    }
-    strcat(comm, " -an");
+        char comm[PATH_MAX + 3] = {0};
+        strncpy(comm, VNETSTAT[VSYSTEMHARDCLASS], (sizeof(comm) - 1));
 
-    FILE *pp;
-    if ((pp = cf_popen(comm, "r", true)) == NULL)
-    {
-        Log(LOG_LEVEL_VERBOSE,
-            "Opening '%s' failed, aborting gathering of network data (monitoring)",
-            comm);
-        return;
-    }
+        if (!FileCanOpen(comm, "r"))
+        {
+            Log(LOG_LEVEL_VERBOSE,
+                "Cannot open '%s', aborting gathering of network data (monitoring)",
+                comm);
+            return;
+        }
+        strcat(comm, " -an");
 
-    Item *in[ATTR] = {0};
-    Item *out[ATTR] = {0};
-    GetNetworkDataFromNetstat(pp, cf_this, in, out);
-    cf_pclose(pp);
+        FILE *pp;
+        if ((pp = cf_popen(comm, "r", true)) == NULL)
+        {
+            Log(LOG_LEVEL_VERBOSE,
+                "Opening '%s' failed, aborting gathering of network data (monitoring)",
+                comm);
+            return;
+        }
+
+        GetNetworkDataFromNetstat(pp, cf_this, in, out);
+        cf_pclose(pp);
+    }
 
     /* Now save the state for ShowState()
        the state is not smaller than the last or at least 40 minutes
        older. This mirrors the persistence of the maxima classes */
     SaveNetworkData(in, out);
 }
 
+#ifdef __linux__
+static inline void SaveSocketInfo(const char *local_addr,
+                                  uint32_t local_port,
+                                  uint32_t remote_port,
+                                  SocketState state,
+                                  SocketType type,
+                                  const char *socket_info,
+                                  double *cf_this,
+                                  Item **in, Item **out)
+{
+    Log(LOG_LEVEL_DEBUG, "Saving socket info '%s:%d:%d [%d, %d]",
+        local_addr, local_port, remote_port, state, type);
+
+    if (state == SOCK_STATE_LISTEN)
+    {
+        char port_str[6];       /* max "65536" (plus '\0') */
```

Comment:
I guess you could... `char port_str[sizeof("65536")];`

```
olehermanse@OHMBP ~ $ cat test.c
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
  return sizeof("65536");
}
olehermanse@OHMBP ~ $ gcc test.c -o test
olehermanse@OHMBP ~ $ ./test 
olehermanse@OHMBP ~ $ echo $?
6
olehermanse@OHMBP ~ $ 
```

No clue whether this works everywhere 🤯  😄

Reply:
_"Relevant"_: https://twitter.com/thingskatedid/status/1295900370309230594

Reply:
We also have `CF_MAX_PORT_LEN` so might as well just use that.